### PR TITLE
improvement: add error context to error creation / normalisation

### PR DESCRIPTION
**Pull request by rellen**
PR for https://github.com/ash-project/ash/issues/345

* add tests for `Ash.Error.to_ash_error` and `Ash.Error.to_error_class`
* add error_context field to `Ash.Error.Exception`
* accumulate error_context in `Ash.Error.to_ash_error` and `Ash.Error.to_error_class`
* use error_context in Ash.Filter error calls
* render context in a 'breadcrumb' in error_messages function call in `Ash.Error`

# Contributor checklist

- [x] Bug fixes include regression tests - N/A
- [x] Chores - N/A
- [x] Documentation changes
- [x] Features include unit/acceptance tests
- [x] Refactoring
- [x] Update dependencies - N/A

### CODE REVIEW

File: lib/ash/error/exception.ex [hunk 1]
```diff
@@ -13,6 +13,7 @@ defmodule Ash.Error.Exception do
                      [
                        :changeset,
                        :query,
+                       error_context: [],
```
Comment (rellen) [inline lib/ash/error/exception.ex:16]: the atom `:context` is used by some errors, which cause a compiler error if we try to use `:context` here.
I've chosen `:error_context`, but not completely happy with that name.  Defs open to suggestions.

File: lib/ash/filter/filter.ex [hunk 2]
```diff
@@ -235,7 +235,10 @@ defmodule Ash.Filter do
         filter
 
       {:error, error} ->
-        raise Ash.Error.to_error_class(error)
+        raise Ash.Error.to_error_class(error,
+                error_context:
+                  "parsing #{inspect(resource)} #{inspect(statement)} #{inspect(aggregates)} #{inspect(calculations)} in #{inspect(context)}"
```
Comment (rellen) [inline lib/ash/filter/filter.ex:240]: I am not sure what actual context we want in some of these call sites.  This is just an example, I doubt this is useful as is.  Will need some guidance on these.

File: lib/ash/filter/filter.ex [hunk 3]
```diff
@@ -872,7 +875,10 @@ defmodule Ash.Filter do
         value
 
       {:error, error} ->
-        raise Ash.Error.to_ash_error(error)
+        raise Ash.Error.to_ash_error(error, nil,
+                error_context:
+                  "parsing #{inspect(base)} #{inspect(addition)} #{inspect(op)} #{inspect(aggregates)} #{inspect(calculations)} in #{inspect(context)}"
```
Comment (rellen) [inline lib/ash/filter/filter.ex:880]: Another placeholder - will need to put in a useful context here.

File: test/error_test.exs [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    test "error message contains error context breadcrumbs" do
+      error1 = Ash.Error.to_ash_error("whoops!", nil, error_context: "some context")
+      error2 = Ash.Error.to_ash_error("whoops, again!!", nil, error_context: "some other context")
+
+      error_class =
+        Ash.Error.to_error_class([error1, error2], error_context: "some higher context")
+
+      error_message = Ash.Error.Unknown.message(error_class)
+
+      assert error_message =~ "Context: some higher context > some context"
```
Comment (rellen) [inline test/error_test.exs:125]: Not sure if these assertions would be considered a bit brittle.
My intent here is that we can write some tests that specify / exemplify the format we want the context to take when rendered in the error message.

File: test/filter/filter_test.exs [hunk 5]
```diff
@@ -566,6 +566,37 @@ defmodule Ash.Test.Filter.FilterTest do
...
+        end)
+
+      [error_context] = err.error_context
+      assert error_context =~ "parsing #Ash.Filter<points =="
+      assert error_context =~ "[bad_field: \"bad field\"]"
+    end
+  end
+
+  describe "parse!" do
+    test "raises an error if the statement is invalid" do
```
Comment (rellen) [inline test/filter/filter_test.exs:585]: might rephrase this test, and add some for the non-throwing version of `parse`

File: test/error_test.exs [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    end
+
+    test "error message contains error context breadcrumbs" do
+      error1 = Ash.Error.to_ash_error("whoops!", nil, error_context: "some context")
+      error2 = Ash.Error.to_ash_error("whoops, again!!", nil, error_context: "some other context")
+
+      error_class =
+        Ash.Error.to_error_class([error1, error2], error_context: "some higher context")
+
+      error_message = Ash.Error.Unknown.message(error_class)
```
Comment (rellen) [inline test/error_test.exs:123]: The error message value for this error is something like:
```Elixir
Unknown Error

Context: some higher context > some context
* whoops!
  (elixir 1.13.4) lib/process.ex:773: Process.info/2
  (ash 2.2.0) lib/ash/error/unknown/unknown_error.ex:5: Ash.Error.Unknown.UnknownError.exception/1
  (ash 2.2.0) lib/ash/error/error.ex:147: Ash.Error.to_ash_error/3
  test/error_test.exs:117: Ash.Test.ErrorTest."test to_error_class error message"/1
  (ex_unit 1.13.4) lib/ex_unit/runner.ex:500: ExUnit.Runner.exec_test/1
  (stdlib 4.1) timer.erl:235: :timer.tc/1
  (ex_unit 1.13.4) lib/ex_unit/runner.ex:451: anonymous fn/4 in ExUnit.Runner.spawn_test_monitor/4
Context: some higher context > some other context
* whoops, again!!
  (elixir 1.13.4) lib/process.ex:773: Process.info/2
  (ash 2.2.0) lib/ash/error/unknown/unknown_error.ex:5: Ash.Error.Unknown.UnknownError.exception/1
  (ash 2.2.0) lib/ash/error/error.ex:147: Ash.Error.to_ash_error/3
  test/error_test.exs:118: Ash.Test.ErrorTest."test to_error_class error message"/1
  (ex_unit 1.13.4) lib/ex_unit/runner.ex:500: ExUnit.Runner.exec_test/1
  (stdlib 4.1) timer.erl:235: :timer.tc/1
  (ex_unit 1.13.4) lib/ex_unit/runner.ex:451: anonymous fn/4 in ExUnit.Runner.spawn_test_monitor/4
```

Comment (rellen) [reply test/error_test.exs:123]: Almost the same as your suggestion in the issue.  I can tweak this more if need be.

File: lib/ash/error/error.ex [hunk 7]
```diff
@@ -182,6 +193,33 @@ defmodule Ash.Error do
...
+
+  defp add_error_context(error, _) do
+    error
+  end
+
+  defp accumulate_error_context(%{errors: [_ | _] = errors} = error, error_context)
+       when is_binary(error_context) do
+    updated_errors =
+      errors
+      |> Enum.map(fn err ->
```
Comment (rellen) [inline lib/ash/error/error.ex:210]: might un-pipeline this

File: test/error_test.exs [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+      assert error_message =~ "Context: some higher context > some context"
+      assert error_message =~ "Context: some higher context > some other context"
+    end
+  end
```
Comment (rellen) [inline test/error_test.exs:128]: could add more tests for `to_error_class` here ad nauseam

File: test/error_test.exs [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+
+  describe "to_ash_error" do
+    test "populates error_context field" do
+      error = Ash.Error.to_ash_error("whoops!", nil, error_context: "some context")
+
+      assert error.error_context == ["some context"]
+    end
```
Comment (rellen) [inline test/error_test.exs:135]: could add more tests for `to_ash_error` here ad nauseam

File: lib/ash/filter/filter.ex (hunk 2 above)
Comment (zachdaniel) [reply lib/ash/filter/filter.ex:240]: I think just the resource and statement is great! :)

File: lib/ash/filter/filter.ex (hunk 3 above)
Comment (zachdaniel) [reply lib/ash/filter/filter.ex:880]: Same comment as above

File: test/error_test.exs (hunk 4 above)
Comment (zachdaniel) [reply test/error_test.exs:125]: I think for now this is a great starting point. If they prove to be brittle for whatever reason we can change them

**Comment by zachdaniel** [comment]
This looks pretty great to me :) I think we could get it merged as is, and then potentially spot-add additional context over time, but I'm also open to the idea of going through and finding additional places to add more context :) As long as the breadcrumbs don't end up too verbose, more context should pretty much always be better.

**Comment by rellen** [comment]
Thanks for the feedback.
I'll clean up the filter context.
I might err on the side of adding a few more call sites and adding some text in, but not necessarily exhaustively.

File: test/error_test.exs (hunk 6 above)
Comment (zachdaniel) [reply test/error_test.exs:123]: looks great!

**Review by zachdaniel — APPROVED** [review]

**Comment by zachdaniel** [comment]
🚀 Thank you for your contribution! 🚀

**Pull request merged**